RESTEasy Reactive Links

gytis opened the pull request:
@FroMage @geoand this is a a draft PR for RESTEasy Reactive Links extension to discuss the API and see if I'm going in a right direction here.

There are three main user facing pieces:

- @RestLink annotation which marks which methods should be considered for link injection (similar to the RESTEasy @LinkResource)
- @InjectRestLinks annotation which indicates a method which response should get link headers injected (it uses RestLinksProvider under the covers). Similar to the RESTEasy RESTServiceDiscovery but instead of injecting the links into an object, it injects them as headers.
- RestLinksProvider interface that has two methods for getting relevant links based on type or object instance. We could use it to generate HAL objects in REST Data Panache or if we'd decide to implement RESTServiceDiscovery from RESTEasy Links.

All JAX-RS methods are scanned in build time when LinksContext is created. In runtime we don't need to do any scans or reflection operations like in RESTEasy Links. It's enough to perform a map lookup to get the link templates and build the actual links.

See [TestResource](https://github.com/quarkusio/quarkus/pull/14249/files#diff-881b95732bfd3e42e3c15ca16c08e1574c72056afa6ab5daa8bd970b62c99d26R20) for an example.

I still need to improve this with more testing, EJB security check and a few other bits. But since I have a roughly functioning prototype I'd like to check in now and see what should be adjusted.

Commit:
RESTEasy Reactive Links

geoand commented:
Great, thanks a lot @gytis! 

I'll have a look tomorrow

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveDeploymentInfoBuildItem.java:
```diff
@@ -0,0 +1,18 @@
+
+import io.quarkus.builder.item.SimpleBuildItem;
+
+public final class ResteasyReactiveDeploymentInfoBuildItem extends SimpleBuildItem {
```
Is this class actually needed? It doesn't seem to be consumed anywhere

gytis replied:
It's consumed in `LinksProcessor`

geoand replied:
Ah OK, I missed that

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/main/java/io/quarkus/resteasy/reactive/links/deployment/PathParameterValueGettersGenerator.java:
```diff
@@ -0,0 +1,76 @@
+import io.quarkus.gizmo.Gizmo;
+import io.quarkus.resteasy.reactive.links.runtime.PathParameterInfo;
+
+public class PathParameterValueGettersGenerator extends ClassVisitor {
```
Let's have some comments on what exactly this does. ASM code is always hard to figure out, so let's at least have comments on what the intended functionality is

gytis replied:
Sure, I'll explain it once I have the extension finalised. But yes, it generates getters for the fields that are needed for path parameter values.

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/main/java/io/quarkus/resteasy/reactive/links/deployment/LinksProcessor.java:
```diff
@@ -0,0 +1,90 @@
+        return new CustomContainerResponseFilterBuildItem(RestLinksResponseFilter.class.getName());
+    }
+
+    @Record(STATIC_INIT) // TODO this doesn't seem to work on reload
```
What doesn't work?

gytis replied:
It doesn't seem to reload the LinksContext. I've just marked this one to look into once we've agreed on the API.

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/runtime/src/main/java/io/quarkus/resteasy/reactive/links/RestLinksResponseFilter.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    private InjectRestLinks getInjectRestLinksAnnotation(ResourceInfo resourceInfo) {
+        if (resourceInfo.getResourceMethod().isAnnotationPresent(InjectRestLinks.class)) {
```
Ideally, we want to avoid this kind of thing as much as possible, as it results in Reflection being performed on the method which otherwisr would not have been performed.

One solution to avoiding the reflection for the initial check, is to downcast to `ResteasyReactiveResourceInfo` and use the `getMethodAnnotationNames` method.

You'll probably have to add a similar method for the class check.

geoand replied:
Another alternative would be to determine the methods that contain the annotations at build time, pass them to a recorder and have your filter look up the information from the info recorded in the recorder.

The general is however that we really want to avoid doing reflection on JAX-RS methods / classes unless it's absolutely necessary

geoand commented:
Thanks a lot for this work @gytis!

I can't really comment on the API as I don't know much about the topic in hand, but I have added a few comments about the code itself

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/test/java/io/quarkus/resteasy/reactive/links/deployment/TestResource.java:
```diff
@@ -0,0 +1,58 @@
+    @GET
+    @Path("/with-links")
+    @Produces(MediaType.APPLICATION_JSON)
+    @RestLink(entityType = TestRecord.class, rel = "list")
```
This should be optional, in which case we can infer that the entity type is `TestRecord` (from the return type `List<TestRecord>` and same for the `rel` which should be `collection`.

gytis replied:
I don't have this implemented yet, but inferring an entity type from a generic is in my todo. So for now this is a workaround.

gytis replied:
As for the rel, this annotation allows to define any rel you want.

FroMage replied:
OK

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/runtime/src/main/java/io/quarkus/resteasy/reactive/links/InjectRestLinks.java:
```diff
@@ -0,0 +1,13 @@
+@Target({ ElementType.TYPE, ElementType.METHOD })
+public @interface InjectRestLinks {
+
+    RestLinkType value() default RestLinkType.TYPE;
```
I wonder what purpose this serves. Can you show me use-cases? The example test could deduce that from the method return type.

gytis replied:
I thought that this would allow user to specify which links does he want to be injected.
E.g. if a `DELETE` operation returns a deleted entity, you don't necessarily want to return the links specific for that entity (self, update, delete) because it doesn't exist any more. But you might still want to return the links for getting all and creating new entities.

FroMage replied:
OK I see, but that feels special-casy.

FroMage reviewed (changes requested):
I added some comments and questions. This looks like limited to `Link` headers ATM, which is fine.
I'm not too sure about the API though, I think it would help if you added examples which we can use as use-cases to verify what it does.
It looks like you're resolving path parameters directly as fields from the resources, which is probably fine, but for resteasy classic (https://docs.jboss.org/resteasy/docs/2.0.0.GA/userguide/html/LinkHeader.html) I only handled IDs. I guess this is more powerful, but requires that the path parameter name matches the entity property.

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/runtime/src/main/java/io/quarkus/resteasy/reactive/links/runtime/PathParameterValueResolver.java:
```diff
@@ -0,0 +1,31 @@
+        for (Method method : instance.getClass().getMethods()) {
+            if (method.getName().equals(pathParameterInfo.getValueGetter())) {
+                try {
+                    return method.invoke(instance);
```
We can get rid of this reflection by generating bytecode that invokes it for a given instance like this:

```java
public class $$Order$Id implements Getter<Order,Long> {
 public static Long get(Order order){
  return order.$$generated_getter_id();
 }
}
```

And you instantiate those getters at static init and can just invoke them without reflection when you need to.

gytis replied:
Nice, I'll add that.

gytis commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/main/java/io/quarkus/resteasy/reactive/links/deployment/PathParameterValueGettersGenerator.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Override
+    public void visitEnd() {
```
It did initially.
I wasn't sure how to check if a method exists from a visitor and I couldn't use an index, because it doesn't get updated. So I've added an extra mapping here https://github.com/quarkusio/quarkus/pull/14249/files/a6de4a1aa44bc8a1102de72835ba98a1355d5c8d#diff-938ce5408bf74bba0f73f4922b625306f3a9deced50ca5afcde36c03d431f3abR49 to make sure that only one getter is generated.

FroMage replied:
I usually collect signatures in `visitMethod`

gytis commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/test/java/io/quarkus/resteasy/reactive/links/deployment/RestLinksInjectionTest.java:
```diff
@@ -0,0 +1,63 @@
+                .getHeaders()
+                .getValues("Link");
+        assertThat(links).containsOnly(
+                Link.fromUriBuilder(UriBuilder.fromUri(testWithLinksUrl).path("/1")).rel("getById").build().toString(),
```
I thought that because we cannot infer all the possible rel values I would simply default it to the method name and allow it to be overwritten with an annotation (`@RestLink(rel="self")`).

gytis commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/test/java/io/quarkus/resteasy/reactive/links/deployment/TestResource.java:
```diff
@@ -0,0 +1,58 @@
+    @Path("/with-links/{id}")
+    @Produces(MediaType.APPLICATION_JSON)
+    @RestLink
+    @InjectRestLinks(RestLinkType.INSTANCE)
```
At the moment by default `TYPE` links are injected i.e. get all and create.

gytis commented:
Thanks for the comments @FroMage. 

> I added some comments and questions. This looks like limited to `Link` headers ATM, which is fine.
For now I've added two things. A `RestLinksProvider ` bean, which returns a collection of links based on type or instance. And a `Link` headers injection, which simply takes the links from the provider and adds link headers. We could inject other types by reusing the same provider. I just didn't want to add too many features without discussing the API.

> I'm not too sure about the API though, I think it would help if you added examples which we can use as use-cases to verify what it does.
This is just a draft to discuss the API. I don't expect it to be final and I'm happy to change it. That's the whole point.

> It looks like you're resolving path parameters directly as fields from the resources, which is probably fine, but for resteasy classic (https://docs.jboss.org/resteasy/docs/2.0.0.GA/userguide/html/LinkHeader.html) I only handled IDs. I guess this is more powerful, but requires that the path parameter name matches the entity property.
Yes I've added a possibility to resolve any entity field. I haven't yet added an override option to the `@RestLink` annotation. With that the path parameter won't need to match the field name. 
Do you think this is too much and I should stick to the ID only?

FroMage commented:
> Yes I've added a possibility to resolve any entity field. I haven't yet added an override option to the @RestLink annotation. With that the path parameter won't need to match the field name.
> Do you think this is too much and I should stick to the ID only?

No, I think it's an interesting approach, and more powerful. Let's try it.

gytis commented:
@FroMage @geoand, sorry for a slight delay in the update but I think I've covered all the comments. However, I still have a couple of points that need to be cleared up.

1. For now I left the `RestLinkType` enum to specify whether type or instance links should be injected. @FroMage had some reservations regarding this point, so I'm open for a debate about that.

2. How sophisticated do we want the Entity type resolution from a generic type to be? For now I only check one level of a single argument parameterised type (e.g. Link, Collection, Set). Does any of you have experience with this? Should we make this deduction more sophisticated? Or maybe we should step back an only check a predefined set of interfaces like RESTEasy Links does?

geoand commented:
My comments have been addressed, the rest I can really comment on since I don't have any experience with links

FroMage reviewed (approved):
LGTM

FroMage commented:
> How sophisticated do we want the Entity type resolution from a generic type to be? For now I only check one level of a single argument parameterised type (e.g. Link, Collection, Set). Does any of you have experience with this? Should we make this deduction more sophisticated? Or maybe we should step back an only check a predefined set of interfaces like RESTEasy Links does?

Just handle collections and special types on a single level, yeah.

gytis commented:
I've just squashed the commits and added the extension file.

geoand commented:
Ah damn... We forgot to merge and now there are conflicts :(

Can you rebase Gytis so we can get this in?

geoand commented:
@Sgitario would you be interested in writing a guide for RESTEasy Reactive links?
The idea being that this module although it's currently only a building block for REST Data Panache, it could however provide a general solution for JAX-RS Paging.

maxandersen commented:
related to #9052 ?

geoand commented:
Related, but we might want something ever higher level

Sgitario commented:
> @Sgitario would you be interested in writing a guide for RESTEasy Reactive links? The idea being that this module although it's currently only a building block for REST Data Panache, it could however provide a general solution for JAX-RS Paging.

I really like the idea. Count on me to do this.

geoand commented:
Thanks a lot!
